### PR TITLE
Update FxCop analyzers

### LINKED ISSUE #17022: Enable FxCop Analyzers
Building on #7467, we should enable FxCop analyers: https://docs.microsoft.com/visualstudio/code-quality/install-fxcop-analyzers. Not only is this recommended by TwC but it allows us to catch things like `catch (Exception ex) { /* do stuff */ throw ex; }` ([CA2200](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2200)).

To note, some of this can be enabled via our existing .editorconfig: https://docs.microsoft.com/dotnet/fundamentals/code-analysis/code-quality-rule-options

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/Azure/azure-sdk-for-net/issues/17022

Big features:

1. Pass cancellation token along
2. Use `stringBuilder.Append(char)` where possible
3. Don't initialize members with the default value
4. Always call `base.Dispose(disposing)`
5. Collection properties should be read-only

### CODE REVIEW

File: eng/Directory.Build.Data.props [hunk 1]
```diff
@@ -25,6 +25,10 @@
       CA2231; <!-- Override Equality operators -->
       CA2225; <!-- Provide alternative to implicit operators -->
       CA1714; <!-- Flags should have plural names -->
+      CA1062; <!-- Public parameter should be checked for null -->
```
Comment (pakrym) [inline eng/Directory.Build.Data.props:28]: Way too many false-positives

File: eng/Directory.Build.Data.props [hunk 2]
```diff
@@ -25,6 +25,10 @@
...
+      CA1031; <!-- Don't catch generic exceptions -->
+      CA2000; <!-- Call dispose on IDisposable objects -->
```
Comment (pakrym) [inline eng/Directory.Build.Data.props:30]: Way too many false-positives

**Review by jsquire — APPROVED** [review]
Event Hubs bits look good to me.

**Review by jamdavi — APPROVED** [review]
FxCopAnalyzer looks good.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs [hunk 3]
```diff
@@ -137,7 +137,7 @@ public virtual TrainingOperation StartTraining(Uri trainingFilesUri, bool useTra
                     ModelName = modelName
                 };
 
-                ResponseWithHeaders<FormRecognizerTrainCustomModelAsyncHeaders> response = ServiceClient.TrainCustomModelAsync(trainRequest);
+                ResponseWithHeaders<FormRecognizerTrainCustomModelAsyncHeaders> response = ServiceClient.TrainCustomModelAsync(trainRequest, cancellationToken);
```
Comment (maririos) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs:140]: 😮😮😮😮 Thank you!

**Review by maririos — APPROVED** [review]
FR and TA LGTM

**Review by heaths — APPROVED** [review]
General, KV, and Search changes LGTM

**Comment by pakrym** [comment]
I'm going to merge so PR doesn't go stale. If you've noticed a mistake please comment and I'll follow up.

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs [hunk 4]
```diff
@@ -19186,7 +19186,9 @@ public partial class BlobServiceProperties
         /// <summary>
         /// The set of CORS rules.
         /// </summary>
+#pragma warning disable CA2227 // Collection properties should be readonly
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs:19189]: @pakrym, This file is auto-generated.  We will have to re-apply this change each time we generate, unless the .NET generator is updated to generate this annotation.

Comment (pakrym) [reply sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs:19189]: Good point, I have a generator update branch at https://github.com/Azure/azure-sdk-for-net/pull/17181

But it seems the generated code is not up-to-date in other cases as well.

**Pull request merged**